Implementation of repr struct alignment RFC 1358.

bitshifter opened the pull request:
The main changes around rustc::ty::Layout::struct:
* Added abi_align field which stores abi alignment before repr align is applied
* align field contains transitive repr alignment
* Added padding vec which stores padding required after fields

The main user of this information is rustc_trans::adt::struct_llfields
which determines the LLVM fields to be used by LLVM, including padding
fields.

A possible future optimisation would be to put the padding Vec in an Option, since it will be unused unless you are using repr align.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @eddyb (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

eddyb commented on src/librustc/ty/layout.rs:
```diff
@@ -523,8 +523,12 @@ pub type FieldPath = Vec<u32>;
     pub align: Align,
 
+    /// Alignment of struct fields, for comparison with LLVM type.
+    pub abi_align: Align,
```
This name is confusing IMO, since each `Align` tracks two values ("preferred" and "ABI-required") - `fields_align` is probably better (if at all needed?).

bitshifter replied:
Naming is hard :) The purpose off this field is to keep track of what LLVM thinks the struct's alignment is. This is used in a few different ways:
* Calculating manual field padding when Rust and LLVM have different ideas about alignment
* Telling alloca alignment of over aligned types (I wanted the be conservative and only specify alloca alignment when they the type is over aligned)
* Sanity checking Rust and LLVM type sizes and alignment in https://github.com/bitshifter/rust/blob/6b9c5d1ff6df1e0d6d181fb5049c3108359b47ee/src/librustc_trans/type_of.rs#L120-L133

Originally I called this field `base_align` which you also suggested in another comment, so perhaps I should go back to that. Other options might be `llvm_align`, `default_align`, `orig_align`.

eddyb replied:
We shouldn't be talking about LLVM in `librustc`, ideally. How about `primitive_align`? That is, the alignment obtained from primitives' alignment, without any user annotations.

bitshifter replied:
`primitive_align` works for me.

eddyb commented on src/librustc/ty/layout.rs:
```diff
@@ -541,6 +545,9 @@ pub struct Struct {
     pub min_size: Size,
+
+    /// Size of padding needed by the struct due to requested repr alignment
+    pub padding: Vec<Size>,
```
Can't this be computed on-demand based on the sizes of the fields and the alignments we can get out of LLVM types? Keep in mind constant enum variant values *already* have to deal with recomputing padding based on the values at hand, I'd rather generalize that algorithm than have this vector.

bitshifter replied:
You comment on this later so I guess you are OK with it?

It could be done on demand, but I thought it would be more overhead as it's needed in a couple of places (https://github.com/bitshifter/rust/blob/6b9c5d1ff6df1e0d6d181fb5049c3108359b47ee/src/librustc_trans/adt.rs#L265-L266 and https://github.com/bitshifter/rust/blob/6b9c5d1ff6df1e0d6d181fb5049c3108359b47ee/src/librustc/ty/layout.rs#L775-L776).

If we keep the vector I was thinking of putting it in an Option, since I think most people won't need `repr(align)` so it would generally be empty.

eddyb replied:
Yeah it's fine if an on-demand iterator happens to use it.

eddyb commented on src/librustc/ty/layout.rs:
```diff
@@ -560,13 +567,27 @@ impl<'a, 'gcx, 'tcx> Struct {
                   repr: &ReprOptions, kind: StructKind,
                   scapegoat: Ty<'gcx>) -> Result<Struct, LayoutError<'gcx>> {
         let packed = repr.packed;
+        let repr_align = repr.align;
```
IMO copying fields out of the struct like that is confusing, both should be referred as `repr.foo`.

eddyb commented on src/librustc/ty/layout.rs:
```diff
@@ -718,6 +767,15 @@ impl<'a, 'gcx, 'tcx> Struct {
+    pub fn memory_index_with_padding_fields(&self, index: usize) -> usize {
+        let index = self.memory_index[index] as usize;
+        self.padding.iter().take(index).fold(index, |acc, &pad|
+            acc + if pad.bytes() > 0 { 1 } else { 0 })
```
Okay, this seems like an okay use of the padding vector.

eddyb commented on src/librustc/ty/layout.rs:
```diff
@@ -816,12 +874,23 @@ impl<'a, 'gcx, 'tcx> Struct {
 #[derive(PartialEq, Eq, Hash, Debug)]
 pub struct Union {
     pub align: Align,
+    pub abi_align: Align,
```
Same, although `field_align` might not work as well here. `member_align` might work for both.

eddyb commented on src/librustc/ty/layout.rs:
```diff
@@ -1513,6 +1603,30 @@ impl<'a, 'gcx, 'tcx> Layout {
     }
+
+    /// Returns alignment before repr alignment is applied
+    pub fn abi_align(&self, dl: &TargetDataLayout) -> Align {
```
Maybe `content_align`? Or, given that it seems to be recursive, `base_align`? (user-specified alignment is still ABI alignment, in the sense that it affects padding, whereas preferred alignment is for allocations, e.g. to speed up SSE).

eddyb commented on src/librustc/ty/mod.rs:
```diff
@@ -1363,6 +1364,7 @@ pub struct ReprOptions {
     pub packed: bool,
     pub simd: bool,
     pub int: Option<attr::IntType>,
+    pub align: Option<u64>,
```
I'd use an `u16` and default it to `0` (we don't support larger alignments anyway - if you want to, you can use a pair of bytes  `Align` instead of just a byte, bringing the max alignment from `2^16` to `2^256`).

eddyb commented on src/librustc_typeck/check/mod.rs:
```diff
@@ -1294,6 +1298,33 @@ pub fn check_simd<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, sp: Span, def_id: DefId
     }
 }
 
+fn has_align<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, root_def_id: DefId, def_id: DefId) -> bool {
```
This looks like it could easily go into infinite recursion.

bitshifter replied:
I was assuming that structs couldn't reference themselves but probably that's assuming too much :).

eddyb replied:
Wait, this only applies to nested `struct`s? That explains why no regressions, *however*, you probably want to handle it so you don't infinitely recurse in the compiler when some other part of the compiler wants to report a non-fatal error about it (see `AdtDef::calculate_sized_constraint_inner`, you probably don't need the cache though, just the stack to check for cycles).

eddyb commented on src/libsyntax/attr.rs:
```diff
@@ -877,10 +886,38 @@ pub fn find_repr_attrs(diagnostic: &Handler, attr: &Attribute) -> Vec<ReprAttr>
+                        if let Ok(align) = u64::from_str(&*value.as_str()) {
+                            if align.is_power_of_two() {
+                                // rustc::ty::layout::Align restricts align to <= 32768
+                                if align <= 32768 {
```
Oh there it is :smile:.

eddyb commented on src/libsyntax/attr.rs:
```diff
@@ -877,10 +886,38 @@ pub fn find_repr_attrs(diagnostic: &Handler, attr: &Attribute) -> Vec<ReprAttr>
+                            }
+                        } else {
+                            span_err!(diagnostic, item.span, E0584,
+                                      "alignment representation hint is not a u64");
```
Oh, we could probably use a single error message for an invalid `#[repr(align)]` (also, "hint" is probably the wrong thing to call it?).

bitshifter replied:
I wasn't sure what the done thing was, lots of specific errors versus one general one. The "hint" part was just from copying the language used in other error messages. I can change it.

bitshifter replied:
Rebased with master and got conflicts on these error numbers, so this all got a bit messed up. Still working on the other feedback.

eddyb commented on src/libsyntax/attr.rs:
```diff
@@ -877,10 +886,38 @@ pub fn find_repr_attrs(diagnostic: &Handler, attr: &Attribute) -> Vec<ReprAttr>
+                if unrecognised {
+                    // Not a word we recognize
+                    span_err!(diagnostic, item.span, E0552,
+                              "unrecognized representation hint");
```
This and the one above it seem like they serve the same function?

bitshifter replied:
Yes, I think it's because there are different code paths for name and named value and for a while I had different error messages. I'll collapse to a single message.

eddyb commented on src/libsyntax/attr.rs:
```diff
@@ -877,10 +886,38 @@ pub fn find_repr_attrs(diagnostic: &Handler, attr: &Attribute) -> Vec<ReprAttr>
                     if let Some(h) = hint {
                         acc.push(h);
                     }
+                } else if let Some((name, value)) = item.name_value_str() {
```
Seems like you want to support only `#[repr(align = "16")]` but not `#[repr(align(16))]`?
@alexcrichton used to think otherwise, but I hope he can be pursued to only allow the latter form, *as it's already implemented* and thus likelier to be stabilized first. In fact, we can stabilize it for specific attributes (like this one) and leave the general case unstable.

alexcrichton replied:
My historical position was that `#[repr(align(16))]` is grammatically invalid (e.g. would simply not parse), but if we've changed the grammar since then then that seems fine by me.

bitshifter replied:
Sure, `[repr(align(16))]` is much nicer. Do you know what else is using that form so I can see how it's implemented?

eddyb replied:
Not sure anything else does that already, but for the record, it's a `MetaItemKind::List` (`.meta_item_list()` on `Attribute`) with a single `NestedMetaItemKind::Literal` element (`.literal()` on `NestedMetaItem`), that's a `LitKind::Int(16, LitIntType::Unsuffixed)`.

bors commented:
:umbrella: The latest upstream changes (presumably #39765) made this pull request unmergeable. Please resolve the merge conflicts.

bitshifter commented:
@eddyb hopefully I've address all your feedback. I made a couple of additional changes:

* Made the Struct padding vector empty unless it's actually used, since the common case won't need it
* added a print-type-sizes test

eddyb commented on src/librustc/ty/layout.rs:
```diff
@@ -546,7 +546,8 @@ pub struct Struct {
-    /// Size of padding needed by the struct due to requested repr alignment
+    /// Size of padding needed by the struct due to requested repr alignment.
+    /// This is created as required and will be empty if the structs fields require no padding.
     pub padding: Vec<Size>,
```
I'm tempted to say that we should have offsets and field sizes, which is more general, and then creating a LLVM type can compute how much padding to use between each two consecutive fields. Check out https://github.com/solson/miri/pull/146#issuecomment-282305202, where I have a potential use for this: removing the implied discriminant field of `enum` variant `Struct`s, which in an `Vec<(field_offset, field_size)>` encoding just be the first field starting e.g. `4` bytes in and thus the LLVM type having a padding field there.

We might also be able to *always* emit padding fields (sometimes zero-sized) and get LLVM to play along.
cc @rust-lang/compiler Anyone think this is terrible idea? IMO it might be better than `memory_index_with_padding_fields` which makes field access `O(n)` instead of `O(1)`.

bitshifter replied:
You mean replacing the Struct offset vec with (field_offset, field_size) tuple and calculating padding on demand when the LLVM type is created? This on it's own wouldn't remove the need for `memory_index_with_padding` though and translating the index to include padding fields would get more expensive. 

I did think about always emitting padding even for primitive aligned types but I wasn't sure what impact that might have on LLVM internals, if that means a lot more work for it to do. Is it possible to emit a zero sized field? Generally I've taken a cautious approach and only changed behaviour if repr align is being used.

BTW, the `memory_index_with_padding_fields` is now only `O(n)` if repr align is being used.

eddyb replied:
Sure, `[0 x i8]` or `{}` (the equivalent of our 0-tuple, `()`) both work to that effect (LLVM ZST).

bitshifter replied:
Ok, so if emit all padding we do that we can make `memory_index_with_padding_fields` always `O(1)` (basically index * 2) and calculate padding on demand.

bitshifter replied:
I did have a couple of unsuccessful attempts at always emitting padding fields. I basically wasn't able to get to the point where I could actually build the compiler, I think it was libcore that I was failing on. I'm not sure how to best debug that. In any case it seemed like it might be a non-trivial change, it seems somewhat independent to the repr align changes, but would make the repr align stuff cleaner. Was there any comment from the compiler team about it?

bitshifter commented on src/librustc/ty/layout.rs:
```diff
@@ -718,6 +771,19 @@ impl<'a, 'gcx, 'tcx> Struct {
+    #[inline]
+    pub fn memory_index_with_padding_fields(&self, index: usize) -> usize {
+        if self.padding.is_empty() {
+            index
```
Whoops. This should be `self.memory_index[index]`.

bors commented:
:umbrella: The latest upstream changes (presumably #40008) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented on src/test/compile-fail/repr-packed-contains-align.rs:
```diff
@@ -0,0 +1,24 @@
+#![feature(attr_literals)]
+#![allow(dead_code)]
+
+#[repr(align(16))]
```
Should these attributes be feature-gated?

nikomatsakis replied:
To be clear: I know there is the `attr_literals` feature gate, but I mean should `align` specifically be feature-gated?

bitshifter replied:
I don't think I have the rustc experience to answer that question :) Happy to add a feature gate if it's needed.

eddyb commented:
Continuing https://github.com/rust-lang/rust/pull/39999#discussion_r103048793: at the compiler teem meeting this week @nikomatsakis seemed fine with the change to always emit padding fields, as long as it had no impact on ABI (and AFAICT, it can't).

@bitshifter Can you push the branch that you couldn't get working somewhere? Maybe I can help.

bitshifter commented:
@eddyb sure, I haven't spent a huge amount of time on it so probably I'm just doing something silly. I've committed my current WIP to https://github.com/bitshifter/rust/commit/00eb42152b256befe12976a3ed8d343c51fbd305. What I'm hitting at the moment is some of my asserts around expected offsets and sizes in `struct_llfields` are being hit when building stage 1 and I haven't worked out why yet. The stride check at the end of that function is commented because it was also failing, which was counter to my expectations too. I haven't worked out yet what struct is being created that triggers these. Hopefully it's not all structs, it's been a week since I tried so no guarantees it's not just totally broken :)

bitshifter commented:
@eddyb I've pushed some changes to that branch so the padding fields are hopefully correct now. I've added a lot of asserts and debug output to `rustc_trans::adt::struct_llfields` and it seemed to be working OK. Now I'm back to getting a segfault when building stage 1 `libcore` (my original attempt was hitting this). The tricky thing is how to debug what's going on. 

Just as I was writing this comment I was trying to attach rust-gdb to the compiler's pid, which seemed to have worked, and I have a stack trace http://pastebin.com/3b8E8Xez. I might take a break from this for now but that trace looks like it should point me in the right direction. If you have any insights or tips for debugging this let me now.

I guess in particular I don't have a good understanding of the different phases of building the compiler. Stage 0 is just downloaded right? Stage 1 appears to start self hosting? At this point it must have built some of the compiler and has started building other libs with that I assume.

bitshifter commented:
More invesitgation, log output for `rustc_trans` where the crash occurs:

```
DEBUG:rustc_trans::mir::statement: trans_statement(statement=_14 = ((_11 as Some).0: <Self as iter::iterator::Iterator>::Item))
DEBUG:rustc_trans::debuginfo::source_loc: set_source_location: src/libcore/iter/iterator.rs:254:13: 254:14
DEBUG:rustc_trans::debuginfo::source_loc: setting debug location to 254 0
DEBUG:rustc_trans::mir::rvalue: trans_rvalue(dest.llval=({ i64, [0 x i8], i32, [4 x i8] }*:  %x = alloca { i64, [0 x i8], i32, [4 x i8] }), rvalue=((_11 as Some).0: <Self as iter::iterator::Iterator>::Item))
DEBUG:rustc_trans::mir::operand: trans_operand(operand=((_11 as Some).0: <Self as iter::iterator::Iterator>::Item))
DEBUG:rustc_trans::mir::operand: trans_consume(lvalue=((_11 as Some).0: <Self as iter::iterator::Iterator>::Item))
DEBUG:rustc_trans::mir::lvalue: trans_lvalue(lvalue=((_11 as Some).0: <Self as iter::iterator::Iterator>::Item))
DEBUG:rustc_trans::mir::lvalue: trans_lvalue(lvalue=(_11 as Some))
DEBUG:rustc_trans::mir::lvalue: trans_lvalue(lvalue=_11)
DEBUG:rustc_trans::monomorphize: apply_param_substs(param_substs=Slice([str::CharIndices]), value=Downcast { adt_def: option::Option, substs: Slice([(usize, char)]), variant_index: 1 })
DEBUG:rustc_trans::mir::lvalue: trans_lvalue(lvalue=(_11 as Some)) => LvalueRef { llval: 0x7f059d15cb78, llextra: 0x0, ty: Downcast { adt_def: option::Option, substs: Slice([(usize, char)]), variant_index: 1 }, alignment: AbiAligned }
DEBUG:rustc_trans::monomorphize: apply_param_substs(param_substs=Slice([str::CharIndices]), value=Ty { ty: <Self as iter::iterator::Iterator>::Item })
DEBUG:rustc_trans::adt: struct_llfields: variant: Struct { align: Align { raw: 51 }, primitive_align: Align { raw: 51 }, packed: false, sized: true, offsets: [Size { raw: 0 }, Size { raw: 8 }], memory_index: [0, 1], min_size: Size { raw: 24 } }
DEBUG:rustc_trans::adt: struct_llfields: 0 ty: u64 min_offset: 0 target_offset: 0
DEBUG:rustc_trans::adt: struct_llfields: 1 ty: (usize, char) pad_bytes: 0 min_offset: 8 target_offset: 8
DEBUG:rustc_trans::adt: struct_llfields: pad_bytes: 0 min_offset: 24 min_size: 24 stride: 24

DEBUG:rustc_trans::mir::lvalue: trans_lvalue(lvalue=((_11 as Some).0: <Self as iter::iterator::Iterator>::Item)) => LvalueRef { llval: 0x7f05a0a17f78, llextra: 0x0, ty: Ty { ty: (usize, char) }, alignment: AbiAligned }
DEBUG:rustc_trans::mir::operand: trans_load: ({ i64, [0 x i8], i32, [4 x i8] }*:  %10 = getelementptr inbounds { i64, [0 x i8], { i64, [0 x i8], i32, [4 x i8] }, [0 x i8] }, { i64, [0 x i8], { i64, [0 x i8], i32, [4 x i8] }, [0 x i8] }* %9, i32 0, i32 2, !dbg !73) @ (usize, char)
```

That source line is in `Iterator::nth` which looks innocuous enough, but it's inlined so I'm not sure what usage is causing the problem.

The crash is inside `LLVMMDNodeInContext`, I need to build that with debug symbols...

eddyb commented:
@bitshifter Oh I haven't gotten a chance to look at it, I still need to bring my own branch to a point where I can rebase yours on it to remove any suspicions of ABI confusion. But if it's debuginfo, that could be missing changes in `rustc_trans::debuginfo`, where it computes metadata for various types.

bitshifter commented:
The pointer for `min` in `Builder::load_range_assert` looks like a garbage value (varies, was 0xff..fe one time). So I'm not sure it's `debuginfo` that's the problem, but getting a bad value out of LLVM somewhere. Do you have any tips for debugging LLVM, it's just a bunch of opaque pointers that I can't introspect.

eddyb commented:
@bitshifter Can you log ([just before this](https://github.com/rust-lang/rust/blob/f573db4f80c75f156df8a743f456bf087ec81dc2/src/librustc_trans/builder.rs#L545)) `Value(ptr)`?
It might not be a pointer to an integer type.

bitshifter commented:
@eddyb log the `LLVMTypeOf`? Got to head out, will take a look when I get back.

eddyb commented:
@bitshifter `debug!("{:?}", Value(ptr))` should show both the type and the instruction.

bitshifter commented:
Ah, helpful. I've pushed a fix which seems to get me past that problem. The next issue is:

```
DEBUG:rustc_trans::back::symbol_export: EXPORTED SYMBOL (local): _ZN4core3num7flt2dec8strategy6dragon12format_exact17h154ec7e8cc7d61b2E (Rust)
DEBUG:rustc_trans::back::write: llvm-optimizing "core.cgu-0"
Invalid operand types for ICmp instruction
  %30 = icmp ne [0 x i8] %29, zeroinitializer, !dbg !21846
Invalid operand types for ICmp instruction
  %55 = icmp ne [0 x i8] %54, zeroinitializer, !dbg !21862
LLVM ERROR: Broken function found, compilation aborted!
```

Don't know if the `EXPORTED SYMBOL` is related or just the previous phase in the log. Looks like an internal assert from LLVM, not sure how to trace the cause.

bitshifter commented:
It seems that `zeroinitializer` is used in constant initialization, maybe that's where this is coming from...

eddyb commented:
Constants are cast so you wouldn't see that. Seems like more field index confusion?

bitshifter commented:
Definitely field index confusion, I just have no idea where it's coming from. The `zeroinitializer` might be a lead? Also there's `!dbg !21846` is it possible to access that debug info from rustc?

eddyb commented:
@bitshifter Ahhh wait a second, do you have LLVM assertions enabled? `--enable-llvm-assertions` to `configure` (or see `src/bootstrap/config.toml.example`).

If you don't, that's why LLVM crashes or finds the problem only in the verifier.

bitshifter commented:
@eddyb according to `config.mk` I have them enabled. Not totally convinced that they really are enabled though, I'll try the `config.toml` approach.

eddyb commented:
@bitshifter It's plausible that LLVM just didn't rebuild since. If I run `./build/*/llvm/bin/opt -version`:
```
LLVM (http://llvm.org/):
  LLVM version 3.9.1
  Optimized build with assertions.
  Default target: x86_64-unknown-linux-gnu
  Host CPU: sandybridge
```

If you want to force a LLVM rebuild, modify `src/rustllvm/llvm-auto-clean-trigger`.

bitshifter commented:
@eddyb thanks, I have a stack trace now! http://pastebin.com/pE7gK4R4. Haven't looked deeper yet, I just find it useful to save these things somewhere :) Looks like something to do with discrinimants.

eddyb commented:
@bitshifter Ahh I was just thinking about this yesterday! So `adt::struct_wrapped_nullable_bitdiscr` will use the "field path" in the `Layout` but there's no doubling of those indices I bet.

bitshifter commented:
@eddyb the stage 1 compiler is actually compiling now! Now to see what tests break...

bitshifter commented:
@eddyb I was looking at adding the zero padding fields in `rustc_trans::mir::constant::build_const_struct` but I wasn't sure if it was necessary. I haven't found exactly how those const fields are indexed yet.

eddyb commented:
@bitshift They're indexed in `O(N)` because of the padding they need has always been non-trivial due to enum variant values not having the enum type - you *could* solve that by adding zero-sized padding.
But it's not needed here and I'd rather merge this sooner than later! Do you want to push the new code?

bitshifter commented:
@eddyb There's a bunch of failing tests right now I need to look at. Some of them may have been related to changes I'd made to the constant stuff which I've backed out again, so will see if the tests pass after that.

bitshifter commented on src/librustc_trans/adt.rs:
```diff
@@ -135,11 +131,9 @@ pub fn finish_type_of<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
                              t: Ty<'tcx>,
                              name: Option<&str>,
-                             sizing: bool,
-                             dst: bool) -> Type {
```
I removed the dst parameter because it was always set to false, seemed like it was no longer used.

eddyb commented on src/librustc_trans/adt.rs:
```diff
@@ -257,34 +251,65 @@ fn union_fill(cx: &CrateContext, size: u64, align: u64) -> Type {
+// Lookup `Struct::memory_index` and double it to account for padding
+pub fn struct_llfields_index(variant: &layout::Struct, index: usize) -> usize {
+    let index = (variant.memory_index[index] as usize) << 1;
+    index
```
No need for a variable here.

bitshifter commented on src/librustc_trans/adt.rs:
```diff
@@ -257,16 +251,68 @@ fn union_fill(cx: &CrateContext, size: u64, align: u64) -> Type {
+}
+
+
+pub fn struct_llfields<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>, field_tys: &Vec<Ty<'tcx>>,
```
At this point it seems like it might be better to have different functions for sizing and non-sizing structs.

eddyb replied:
I have a branch which completely kills "sizing types", although I think I'll wait for this to be merged first.

eddyb commented:
@bitshifter Looks like a lot of segfaults in compile-fail tests (i.e. the compiler can't output errors).

bitshifter commented:
It seems that these compile-fail tests pass in stage 1 but fail in stage 2 with an internal compiler error

```thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:323
stack backtrace:
   0:     0x7f0afe195a63 - std::sys::imp::backtrace::tracing::imp::unwind_backtrace::h84cf1b4471513b76
                               at /data/rust/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1:     0x7f0afe18f487 - std::sys_common::backtrace::_print::hc1fb9e40cbd49ba2
                               at /data/rust/src/libstd/sys_common/backtrace.rs:71
   2:     0x7f0afe1a48fd - std::panicking::default_hook::{{closure}}::h6c995967d73dafc7
                               at /data/rust/src/libstd/sys_common/backtrace.rs:60
                               at /data/rust/src/libstd/panicking.rs:355
   3:     0x7f0afe1a445f - std::panicking::default_hook::hbce7d419506cf292
                               at /data/rust/src/libstd/panicking.rs:365
   4:     0x7f0afe1a4ef6 - std::panicking::rust_panic_with_hook::h0908b2962a1212f2
                               at /data/rust/src/libstd/panicking.rs:549
   5:     0x7f0afe1a4d84 - std::panicking::begin_panic::h28f23e30564aa045
                               at /data/rust/src/libstd/panicking.rs:511
   6:     0x7f0afe1a4cb9 - std::panicking::begin_panic_fmt::h119da7f4dd3dfd80
                               at /data/rust/src/libstd/panicking.rs:495
   7:     0x7f0afe1a4c47 - rust_begin_unwind
                               at /data/rust/src/libstd/panicking.rs:471
   8:     0x7f0afe1eb88d - core::panicking::panic_fmt::h0a6e703df3c39315
                               at /data/rust/src/libcore/panicking.rs:69
   9:     0x7f0afe1eb7c4 - core::panicking::panic::hcf9e3d59a8633672
                               at /data/rust/src/libcore/panicking.rs:49
  10:     0x7f0afd5ee1db - rustc_resolve::resolve_imports::ImportResolver::finalize_imports::h9e620209e106a194
                               at /data/rust/src/libcore/macros.rs:21
                               at /data/rust/src/librustc_resolve/resolve_imports.rs:693
                               at /data/rust/src/librustc_resolve/resolve_imports.rs:462
  11:     0x7f0afd5f685c - rustc_resolve::Resolver::resolve_crate::h6b4afa7bcd68dfdd
                               at /data/rust/src/librustc_resolve/lib.rs:1415
  12:     0x7f0afe52126c - rustc_driver::driver::phase_2_configure_and_expand::h44be10961c7d21bb
                               at /data/rust/src/librustc_driver/driver.rs:764
                               at /data/rust/src/librustc/util/common.rs:48
                               at /data/rust/src/librustc_driver/driver.rs:763
  13:     0x7f0afe518196 - rustc_driver::driver::compile_input::h0ca3d134ab854074
                               at /data/rust/src/librustc_driver/driver.rs:114
  14:     0x7f0afe55f711 - rustc_driver::run_compiler::h1760df6c109ed010
                               at /data/rust/src/librustc_driver/lib.rs:221
  15:     0x7f0afe48d291 - std::panicking::try::do_call::h77393471dabbbc8b
                               at /data/rust/src/librustc_driver/lib.rs:1130
                               at /data/rust/src/librustc_driver/lib.rs:137
                               at /data/rust/src/librustc_driver/lib.rs:1064
                               at /data/rust/src/libstd/panic.rs:296
                               at /data/rust/src/libstd/panicking.rs:454
  16:     0x7f0afe1ae57a - __rust_maybe_catch_panic
                               at /data/rust/src/libpanic_unwind/lib.rs:98
  17:     0x7f0afe4a9efb - <F as alloc::boxed::FnBox<A>>::call_box::h4f527d7492852135
                               at /data/rust/src/libstd/panicking.rs:433
                               at /data/rust/src/libstd/panic.rs:361
                               at /data/rust/src/libstd/thread/mod.rs:360
                               at /data/rust/src/liballoc/boxed.rs:640
  18:     0x7f0afe1a35e4 - std::sys::imp::thread::Thread::new::thread_start::hcc5fbed4d4c6ad91
                               at /data/rust/src/liballoc/boxed.rs:650
                               at /data/rust/src/libstd/sys_common/thread.rs:21
                               at /data/rust/src/libstd/sys/unix/thread.rs:84
  19:     0x7f0af61aa6b9 - start_thread
  20:     0x7f0afde4782c - clone
  21:                0x0 - <unknown>
```

eddyb commented:
@bitshifter Yeah, miscompiled compiler. Last time I had something like this happen, some of the constants were broken *shrug*.

bitshifter commented:
@eddyb I did see a stage 1 error in the tests for `libserialize/json.rs` in this test:

```
    #[test]
    fn test_hashmap_with_enum_key() {
        use std::collections::HashMap;
        use json;
        #[derive(RustcEncodable, Eq, Hash, PartialEq, RustcDecodable, Debug)]
        enum Enum {
            Foo,
            #[allow(dead_code)]
            Bar,
        }
        let mut map = HashMap::new();
        map.insert(Enum::Foo, 0);
        let result = json::encode(&map).unwrap();
        assert_eq!(&result[..], r#"{"Foo":0}"#);
        let decoded: HashMap<Enum, _> = json::decode(&result).unwrap();
        assert_eq!(map, decoded);
    }
```

The `assert_eq` failed. So that's something to look at.

bitshifter commented:
@eddyb It was probably a bit premature merging this back to the PR :(

Anyway, I've found a handful of failures in run-pass tests:

```
run-pass/dst-field-align.rs
run-pass/issue-26905.rs
run-pass/issue28498-ugeh-with-lifetime-param.rs
run-pass/issue28498-ugeh-with-trait-bound.rs
run-pass/minmax-stability-issue-23687.rs
run-pass/vector-sort-panic-safe.rs
```

Hopefully they're actually just one or two bugs. In the couple that I've looked at there does seem to be a pattern of nested generic types (DST in Rust terminology?) being accessed. `issue-26905.rs` is actually asserting inside LLVM so it's probably the easiest one to start with. The back trace is http://pastebin.com/vxqtXPN2. Something to do with `rustc_trans::mir::operand::{{impl}}::pack_if_pair`. I'm not familiar with this area of code yet so if you have any insights let me know.

eddyb commented:
@bitshifter I've finally managed to get my ABI stuff working (#40658), although I think I was mistaken in believing some of it might be needed to make this PR work correctly, except for maybe AArch64 HFA/HVA.

bitshifter commented:
I *think* I'm down to 1 broken codegen test, unless others pop up once it's fixed. Maybe this PR will be ready soon.

eddyb commented:
@bitshifter Oh this last one looks like it needs my PR to compute ABI correctly, guess I didn't wadte my time.

eddyb commented:
@bitshifter Btw, before you even attempt rebasing, I *strongly* suggest squashing the commits that change what a previous commit did, i.e. these ones:
* Rename Struct abi_align and use u16 for repr align.
* Switched repr attr from named value to a literal
* Allocate struct padding vector on demand.
* Commenting pass
* Actually return the adjusted memory index.
* Always emit padding fields for LLVM structs.
* Remove unnecessary variable
* Fixes for aligned struct test.

That way, you won't have to redo the same fix half a dozen times during the rebase.

eddyb commented:
@bitshifter Just confirmed that rebasing your changes on top of #40658 passes all the codegen tests!

bitshifter commented:
@eddyb Great! When do you think your PR will make it into master? I'll look at flattening those commits now. Would it make sense to just flatten them all down to one?

eddyb commented:
@bitshifter That's fine, as long as there's nothing you want to keep separate for some reason.

eddyb commented on src/libsyntax/attr.rs:
```diff
@@ -866,20 +885,38 @@ pub fn find_repr_attrs(diagnostic: &Handler, attr: &Attribute) -> Vec<ReprAttr>
+                        }
+                        if !valid_align {
+                            span_err!(diagnostic, item.span, E0589,
+                                      "align representation must be a u16 power of two");
```
"must be a power of two less than 32768" seems more accurate.

eddyb replied:
It's also possible to have a different message for every kind of problem (feeding into the same `span_err!`).

bitshifter replied:
What kind of different problems are you anticipating here? I did used to have separate messages but collapsed it down to one.

eddyb replied:
I'm seeing: "not an unsuffixed integer", "not a power of two", "larger than `32768`".
I think each should have a separate message *but not separate error code* (i.e. all are "invalid `repr(align)`).

bitshifter replied:
Ahh, I didn't realise you could have different messages for the same error code :)

eddyb replied:
It's formatting so you can have `"common part: {}", message`.

bitshifter replied:
I've made all of these changes. I will attempt to rebase with master again but if I'm unable to build stage one I'll just leave it until #40658 is in.

eddyb commented on src/libsyntax/diagnostic_list.rs:
```diff
@@ -287,10 +287,10 @@ register_diagnostics! {
     E0556, // malformed feature, expected just one word
     E0557, // feature has been removed
     E0584, // file for module `..` found at both .. and ..
+    E0589, // align representation must be a u16 power of two
```
Message would have to be updated here too.

eddyb commented on src/libsyntax/lib.rs:
```diff
@@ -35,6 +35,7 @@
 #![feature(specialization)]
 #![feature(i128_type)]
 
+extern crate core;
```
What's this for?

bors commented:
:umbrella: The latest upstream changes (presumably #39628) made this pull request unmergeable. Please resolve the merge conflicts.

bitshifter commented:
@eddyb I did a test rebase against your branch + latest master. I didn't do a full build/and test be it looks good, the stage 1 build assert I was hitting with master is gone and codegen tests all pass.

bitshifter commented:
@eddyb : I spoke too soon. A full build fails to compile stage 2 rustc (result is the same with and without your change). The call stack is https://pastebin.com/w9LKQ56u. I'm still investigating, I think it's something to do with `Mir::split_args`.

eddyb commented:
@bitshifter #40658 just got merged, I'm going to try to rebase this PR locally to see what might be wrong.

bors commented:
:umbrella: The latest upstream changes (presumably #40658) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented:
@bitshifter Stage 2 just compiled fine for me on this branch (after a rebase on master).

bitshifter commented:
The following tests are failing:

 ```
[ui] ui/print_type_sizes/nullable.rs
[ui] ui/print_type_sizes/packed.rs
[ui] ui/print_type_sizes/padding.rs
[ui] ui/print_type_sizes/repr_align.rs # fixed
```

These have failed before but I can't recall what I did to fix them :)

eddyb commented:
@bitshifter You have to update their output I guess? Assuming the output is correct.

bitshifter commented:
On vacation for a week so probably won't have time to look at these
immediately

On 10 Apr 2017 11:08 p.m., "Eduard-Mihai Burtescu" <notifications@github.com>
wrote:

> @bitshifter <https://github.com/bitshifter> You have to update their
> output I guess? Assuming the output is correct.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/39999#issuecomment-292919160>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAISFIaBSLVxGjpSAKnQpxBgSKzYHB6tks5rug2_gaJpZM4MHQVz>
> .
>

bors commented:
:umbrella: The latest upstream changes (presumably #41237) made this pull request unmergeable. Please resolve the merge conflicts.

arielb1 commented:
Friendly ping to @bitshifter.

bitshifter commented:
Hi! I'm back from holiday. Current status is stage 1 doesn't compile after
rebase from about a week ago, there's an llvm assert. I'm still
investigating the problem and holding off pushing until it compiles. I'm
hoping I'll be able to finish this off on the weekend. It was really close
before master broke things again.

On 18 Apr 2017 6:55 p.m., "Ariel Ben-Yehuda" <notifications@github.com>
wrote:

> Friendly ping to @bitshifter <https://github.com/bitshifter>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/39999#issuecomment-294735780>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAISFLuxtb-FPQqaz0RoQrTcGSwlx46Sks5rxHqCgaJpZM4MHQVz>
> .
>

eddyb commented:
@bitshifter It's probably due to field reordering being enabled, perhaps you didn't take into account?

eddyb commented on src/librustc_trans/mir/rvalue.rs:
```diff
@@ -143,11 +145,7 @@ impl<'a, 'tcx> MirContext<'a, 'tcx> {
-                                    i
-                                };
+                                let i = get_memory_index(i);
                                 let dest = bcx.gepi(dest.llval, &[0, i]);
```
Hmpf, this should probably just be using `trans_field_ptr`.

eddyb commented on src/librustc_trans/mir/mod.rs:
```diff
@@ -384,9 +385,13 @@ fn arg_local_refs<'a, 'tcx>(bcx: &Builder<'a, 'tcx>,
             let lvalue = LvalueRef::alloca(bcx, arg_ty, &format!("arg{}", arg_index));
             for (i, &tupled_arg_ty) in tupled_arg_tys.iter().enumerate() {
-                let dst = bcx.struct_gep(lvalue.llval, i);
+                let dst = bcx.struct_gep(lvalue.llval, adt::struct_llfields_index(variant, i));
```
I would have to see the rebased version to tell you if what you did here is correct.

eddyb replied:
The rebased version looks fine 😞

bitshifter commented:
I pushed a WIP backup to
https://github.com/bitshifter/rust/tree/struct_align_rebase?files=1

On 18 Apr 2017 7:53 p.m., "Eduard-Mihai Burtescu" <notifications@github.com>
wrote:

> *@eddyb* commented on this pull request.
> ------------------------------
>
> In src/librustc_trans/mir/mod.rs
> <https://github.com/rust-lang/rust/pull/39999#discussion_r111913613>:
>
> >              let lvalue = LvalueRef::alloca(bcx, arg_ty, &format!("arg{}", arg_index));
>              for (i, &tupled_arg_ty) in tupled_arg_tys.iter().enumerate() {
> -                let dst = bcx.struct_gep(lvalue.llval, i);
> +                let dst = bcx.struct_gep(lvalue.llval, adt::struct_llfields_index(variant, i));
>
> I would have to see the rebased version to tell you if what you did here
> is correct.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/39999#pullrequestreview-33187525>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAISFCefJXKWRFLLvvG5JF7jxuDEHlxqks5rxIgMgaJpZM4MHQVz>
> .
>

eddyb commented:
@bitshifter This is the interesting view: https://github.com/rust-lang/rust/compare/master...bitshifter:struct_align_rebase

bitshifter commented:
When I was looking into it a week ago it was a bit cast from a zero pad
field,I just haven't worked out where it's coming from yet.

On 18 Apr 2017 8:09 p.m., "Eduard-Mihai Burtescu" <notifications@github.com>
wrote:

> *@eddyb* commented on this pull request.
> ------------------------------
>
> In src/librustc_trans/mir/mod.rs
> <https://github.com/rust-lang/rust/pull/39999#discussion_r111916832>:
>
> >              let lvalue = LvalueRef::alloca(bcx, arg_ty, &format!("arg{}", arg_index));
>              for (i, &tupled_arg_ty) in tupled_arg_tys.iter().enumerate() {
> -                let dst = bcx.struct_gep(lvalue.llval, i);
> +                let dst = bcx.struct_gep(lvalue.llval, adt::struct_llfields_index(variant, i));
>
> The rebased version looks fine 😞
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/39999#discussion_r111916832>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAISFKeUvfSfl50k6comCkfFV-dlmB7Wks5rxIuWgaJpZM4MHQVz>
> .
>

bitshifter commented:
This is the failing callstack https://pastebin.com/hea20yZ4

I don't think I've really made much progress in working out what the cause is sadly.

eddyb commented:
@bitshifter Can you get `RUST_LOG=rustc_trans` output for the last function (you should be able to see where one function starts and the next one begins, the whole output might be huge)?
I think an input to a (checked) multiplication is one of those padding fields, but it's unclear what loaded it.

bitshifter commented:
@eddyb yeah I think that is the case, a zero pad field is being used as input to the checked mul. I've pasted a bit of log to https://pastebin.com/Zq8iHCK1. Hopefully that's enough to see what's going on.

eddyb commented:
@bitshifter It's trying to load the third (`hex_digit_idx`) field of `char::EscapeUnicode`.
`memory_index: [1, 2, 0]` ends up in `0, 0` GEPi (which seems fine) but the resulting type is `[0 x i8]*`?!

eddyb commented on src/librustc_trans/adt.rs:
```diff
@@ -247,15 +242,58 @@ fn union_fill(cx: &CrateContext, size: u64, align: u64) -> Type {
+    let field_iter = variant.field_index_by_increasing_offset().map(|i| {
+        (i, field_tys[i as usize], variant.offsets[i as usize].bytes()) });
+    for (index, ty, target_offset) in field_iter {
+        if index != 0 {
```
This is wrong with field reordering, you should just use a local boolean `first` or something.

bitshifter replied:
Ahh, good find. I'll have a look when I get home from work.

bors commented:
:umbrella: The latest upstream changes (presumably #41411) made this pull request unmergeable. Please resolve the merge conflicts.

frewsxcv commented:
FYI, my PR https://github.com/rust-lang/rust/pull/41295 just landed which made a couple organizational changes to the Unstable Book. Namely, features are now divided into 'language features' and 'library features' which are different directories, so you'll unfortunately have to address the conflicts. You should be able to just run `./x.py test src/tools/tidy` to see what needs to be changed after rebasing/merging in master. Sorry for the inconvenience and if you need any help with that let me know! I can address the merge conflicts for you if you want.

bitshifter commented:
Conflicts resolved and all tests passing locally. Hopefully travis agrees.

bitshifter commented:
@eddyb it looks like travis passed! I think I addressed your review change requests, a bit has changed since then though so you'll probably want to have another pass over it.

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit c219cdf has been approved by `eddyb`

<!-- @bors r=eddyb c219cdfa11b0bb23f1dabbef9415ffb94f10386a -->

nagisa commented:
Is intrafield padding used everywhere with this PR?

On Apr 21, 2017 8:07 AM, "bors" <notifications@github.com> wrote:

> 📌 Commit c219cdf
> <https://github.com/rust-lang/rust/commit/c219cdfa11b0bb23f1dabbef9415ffb94f10386a>
> has been approved by eddyb
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/39999#issuecomment-296067626>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AApc0phsJh9ylLFXypWbXoMQmrEQLEqaks5ryDmYgaJpZM4MHQVz>
> .
>

bitshifter commented:
@nagisa yes, that was @eddyb's request. From my perspective it's good because the code gen for repr align types is the same as without, which means it will be well tested.

bors commented:
:hourglass: Testing commit c219cdf with merge 6c7b5d1...

nagisa commented:
I was asking about this because i remember LLVM having worse codegen with
implied padding compared to codegen with explicit padding (which is what
this PR does). This PR might result in faster code overall :)

On Apr 21, 2017 08:32, "Cameron Hart" <notifications@github.com> wrote:

> @nagisa <https://github.com/nagisa> yes, that was @eddyb
> <https://github.com/eddyb>'s request. From my perspective it's good
> because the code gen for repr align types is the same as without, which
> means it will be well tested.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/39999#issuecomment-296074598>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AApc0pk_5wKrB29oaesDdBb7iIwFps6Cks5ryD-CgaJpZM4MHQVz>
> .
>

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3002)

bitshifter commented:
That would be a nice side effect :)

On 21 April 2017 at 16:56, Simonas Kazlauskas <notifications@github.com>
wrote:

> I was asking about this because i remember LLVM having worse codegen with
> implied padding compared to codegen with explicit padding (which is what
> this PR does). This PR might result in faster code overall :)
>
> On Apr 21, 2017 08:32, "Cameron Hart" <notifications@github.com> wrote:
>
> > @nagisa <https://github.com/nagisa> yes, that was @eddyb
> > <https://github.com/eddyb>'s request. From my perspective it's good
> > because the code gen for repr align types is the same as without, which
> > means it will be well tested.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/rust-lang/rust/pull/39999#issuecomment-296074598>,
> or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/AApc0pk_
> 5wKrB29oaesDdBb7iIwFps6Cks5ryD-CgaJpZM4MHQVz>
>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/39999#issuecomment-296098253>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAISFMNOF-p-v6JNT470vr8bBMtJLDLMks5ryFMmgaJpZM4MHQVz>
> .
>

bitshifter commented:
failed to build LLVM on MSYS...

On 21 April 2017 at 16:59, bors <notifications@github.com> wrote:

> 💔 Test failed - status-appveyor
> <https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3002>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/39999#issuecomment-296099156>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAISFA_7xyzhyloES60DI1SiPGyTu3yqks5ryFPagaJpZM4MHQVz>
> .
>

eddyb commented:
@bors retry

bors commented:
:hourglass: Testing commit c219cdf with merge a032921...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/224288684)

eddyb commented:
@bitshifter [Failure on ARM](https://s3.amazonaws.com/archive.travis-ci.org/jobs/224288687/log.txt).

bitshifter commented:
@eddyb It's failing in `rustc_trans::adt::generic_type_of`in this assert for `Layout:General`:

```
assert_eq!(machine::llalign_of_min(cx, variant_fill), align as u32);
```

Any idea why this would be different on ARM?

I'm not sure how to go about debugging this, I don't have access to an ARM dev environment.

bitshifter commented:
@eddyb Maybe something to do with `rustc_trans::adt::union_fill`. I think in this instance, the `else` branch of the below code would be hit

```
    if let Some(ity) = layout::Integer::for_abi_align(cx, layout_align) {
        Type::array(&Type::from_integer(cx, ity), align_units)
    } else {
        Type::array(&Type::vector(&Type::i32(cx), align/4),
                    align_units)
    }
```
Perhaps the `Type::vector` behaves differently on this arch, or it doesn't support wide enough vectors?

eddyb commented:
@bitshifter We should not be emitting vector types, they're useless for this 😞

bitshifter commented:
@eddyb I guess this code is trying to not add extra fields for padding. I wonder what the motivation for using a vector here was. Maybe I could just remove the vector branch.

bitshifter commented:
I think I understand what this code is trying to do - to create space for the tagged enum variants using a type that has the same alignment as the max enum variant alignment. I'm not sure why this is necessary though? This code is really old, I think the vector code path was added 3 years ago in https://github.com/rust-lang/rust/commit/8658a433f2ad46e5834d2168a73df9cae88c29e5. The layout should store the maximum variant's alignment and I think this is the alignment that gets used. That's the case for repr aligned variants anyway.

Can you see any problem with just emitting i8's for the enum contents? I'm trying it now, so far it appears to work, but perhaps there are arch differences with this approach.

eddyb commented:
@bitshifter It's just a belief that larger elements are better, and I think vectors were used to get 16-byte elements but I don't know more. You can only really do it for `primitive_align` though, not the possibly-user-specified `align`.

bitshifter commented:
@eddyb If the alignment doesn't matter, perhaps the easiest thing to do is just remove that assert.

eddyb commented:
@bitshifter Hmm now that I think more about it, I believe the problem is comparing an alignment from LLVM with `align` instead of `primitive_align`. LLVM *can't* return anything *other* than `primitive_align`.

bitshifter commented:
@eddyb that would make sense, except for some reason this assert isn't firing on x86_64. I really don't understand why not.

eddyb commented:
@bitshifter Just give an `enum` the alignment of 32 or more. x64 has 16-byte alignment SSE.

bitshifter commented:
@eddyb I did try that, it still didn't assert... I'll try changing the assert to primitive align anyway and see what happens.

eddyb commented:
@bitshifter Sorry, I meant something like `#[repr(align(64))] struct Foo(u8);` and then trying to use `Option<Foo>`.

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 946f8e6 has been approved by `eddyb`

<!-- @bors r=eddyb 946f8e6a59242a8112c6983d1336fef54bc55b9a -->

bitshifter commented:
@eddyb weirdly, the 64 bit aligned `Option<Foo>` doesn't assert. I added some print statements and LLVM is reporting the alignment as 64. I think perhaps it's because of this https://github.com/rust-lang/llvm/blob/2e951c3ae354bcbd2e50b30798e232949a926b75/lib/IR/DataLayout.cpp#L494. LLVM must let you create arbitrary vector types, I think it's not a problem because there's no code generated that tries to access the data as that type. Still not sure why this would be different on ARM though.

Anyway, I don't see any problem with the change I made, so hopefully it passes homu tests. I don't think alignment should be important here, as it should be handled by the `Layout`.

eddyb commented:
@bitshifter I am dumb, 64 bytes is AVX512. I should've said 1024 bytes or something.

bitshifter commented:
I did try 1024 yesterday with the enum sample I pasted on an earlier comment

On 22 Apr 2017 3:46 p.m., "Eduard-Mihai Burtescu" <notifications@github.com>
wrote:

> @bitshifter <https://github.com/bitshifter> I am dumb, 64 bytes is
> AVX512. I should've said 1024 bytes or something.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/39999#issuecomment-296349970>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAISFMhQH2IUEbnUTE1NdZ-Ny4Iz9prwks5ryZQngaJpZM4MHQVz>
> .
>

bors commented:
:hourglass: Testing commit 946f8e6 with merge 6d841da...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3018), [status-travis](https://travis-ci.org/rust-lang/rust/builds/224654508)
Approved by: eddyb
Pushing 6d841da4a0d7629f826117f99052e3d4a7997a7e to master...

bitshifter commented:
I should have edited the PR description before it got merged, it's pretty out of date :)

There's some follow up things that should probably be done in the future:
* write a codegen test - the `test/run-pass/align-struct.rs` doesn't catch everything - `alloca`'s missing the alignment can just work depended on how they're used (e.g. address might happen to have correct alignment, only fails if usage requires alignment like vector instruction load/stores)
* specify alignment for over aligned types to stores and loads - this is an optimisation. it's always safe to under specify alignment, it's just slower, but over specifying is undefined behaviour so care is required. I thought it would be good to see how `repr(align)` is being used before tackling this.

What would the process be for tracking these, create some new issues?

eddyb commented:
@bitshifter I'd put them in the tracking issue, if there is one for `#[repr(align)]`.
